src/transform/src/analysis/equivalences.rs

```diff
@@ -525,9 +520,18 @@ impl EquivalenceClasses {
         self.classes.extend(to_add);
 
         // Tidy up classes, restore representative.
+        // Specifically, we want to remove literal errors before restoring the equvialence class structure.
         self.tidy();
 
-        stable
+        // The state is stable if every expression is present in `remap` with the same representative, and their sizes are the same.
```

Comment:
Can you explain the advantage of this new stability check compared to the old code that simply set `stable = false` when a change happens? `minimize_once` does a lot of stuff, it seems non-trivial to me to verify that this new check achieves the same as the old code. (Also, the old code is, I guess, faster.)

Reply:
The old check required, among other things, defensive copies of every expression, because `MRE::reduce` does not produce a `bool` result indicating whether a change happened. So we were already cloning everything and doing equality checks on them. The PR moves that cloning into the formation of `remap`, independently needed for efficient operation, but since we now have a copy of the ground truth we started with, we use it rather than make defensive copies and bespoke change tracking.

> it seems non-trivial to me to verify that this new check achieves the same as the old code.

100%. The new check is more likely to achieve *the correct answer*, and the complexity is in verifying that the old code even did that in the first place.

Reply:
Hmm, eliminating the defensive copies is a good point!

When you say that the new code's answer is more correct, you mean that the old code might have sometimes set `stable = false` even when we were actually already stable, due to making some change that was not actually a meaningful change?

Reply:
Ok, the new check seems correct, indeed.

Reply:
> When you say that the new code's answer is more correct ...

Imo it's not about the answer as much as the process. The new code almost directly confirms that the contents of `classes` at the start and at the end are identical. The old code had bespoke change tracking that relied on some amount of care in the implementation. It will be harder for the new code to produce a wrong answer going forward, but the old code could pretty easily do that.